### DISCUSSION TITLE
Specific language shell

### DISCUSSION THREAD

**Discussion by SadE54** [tui-cs/Terminal.Gui#1555] [General]
Hello! 

I need to develop a shell like application, that is able to provide auto-completion suggestion in a combo list like way. The suggestions are provided dynamically according to the lexer/parser state . Is this is something 'easily' doable by gui.cs and would you recommend it for such a feature?

**Comment by SadE54** [comment]
I forgot to mention the need to syntax highlight some content of the prompt line  as it's being typed :)

**Comment by tznind** [comment]
`TextView` may fit the bill for you.  It supports both [syntax highlighting](https://github.com/migueldeicaza/gui.cs/discussions/1553) and [autocomplete](https://github.com/migueldeicaza/gui.cs/pull/1406).

However TextView is designed to work more like a text editor than a shell window so some work would be needed to get the effect you want.  If you want things like window frames and other views (tabs panes, tables, graphs etc) then this is definetly a good fit for Terminal.Gui 👍 

One thing to note (since you metnioned shell) is that if you are spawning subprocesses you need to redirect the console input/output streams (see https://github.com/migueldeicaza/gui.cs/issues/1547#issuecomment-995259708 ) so as not to corrupt the console I/O stream to Terminal.Gui (in linux running with ncurses driver).

**Comment by SadE54** [comment]
Thanks for the answer ! 
[https://github.com/migueldeicaza/gui.cs/pull/1406](url) is exactly what I want to achieve. I told about shell , but it's more like the sql language in the example. My purpose is to edit live some specific xml files with commands like get/set property.

**Reply by tznind** [reply]
That's sounds cool and well within the remit of the controls.  You might want to start by copying and adjusting this scenario (the sql one) https://github.com/migueldeicaza/gui.cs/blob/main/UICatalog/Scenarios/SyntaxHighlighting.cs
